

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>General usage</title>
    
    <link rel="stylesheet" href="../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.16.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/bedtools.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="The BEDTools suite" href="bedtools-suite.html" />
    <link rel="prev" title="Quick start" href="quick-start.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="bedtools-suite.html" title="The BEDTools suite"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="quick-start.html" title="Quick start"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="general-usage">
<h1>General usage<a class="headerlink" href="#general-usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="supported-file-formats">
<h2>4.1 Supported file formats<a class="headerlink" href="#supported-file-formats" title="Permalink to this headline">¶</a></h2>
<div class="section" id="bed-format">
<h3>4.1.1 BED format<a class="headerlink" href="#bed-format" title="Permalink to this headline">¶</a></h3>
<p>As described on the UCSC Genome Browser website (see link below), the BED format is a concise and
flexible way to represent genomic features and annotations. The BED format description supports up to
12 columns, but only the first 3 are required for the UCSC browser, the Galaxy browser and for
BEDTools. BEDTools allows one to use the &#8220;BED12&#8221; format (that is, all 12 fields listed below).
However, only intersectBed, coverageBed, genomeCoverageBed, and bamToBed will obey the BED12
&#8220;blocks&#8221; when computing overlaps, etc., via the <strong>&#8220;-split&#8221;</strong> option. For all other tools, the last six columns
are not used for any comparisons by the BEDTools. Instead, they will use the entire span (start to end)
of the BED12 entry to perform any relevant feature comparisons. The last six columns will be reported
in the output of all comparisons.</p>
<p>The file description below is modified from: <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat#format1">http://genome.ucsc.edu/FAQ/FAQformat#format1</a>.</p>
<ol class="arabic simple">
<li><strong>chrom</strong> - The name of the chromosome on which the genome feature exists.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;chr1&#8221;, &#8220;III&#8221;, &#8220;myChrom&#8221;, &#8220;contig1112.23&#8221;.</li>
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><strong>start</strong> - The zero-based starting position of the feature in the chromosome.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The first base in a chromosome is numbered 0</em>.</li>
<li><em>The start position in each BED feature is therefore interpreted to be 1 greater than the start position listed in the feature. For example, start=9, end=20 is interpreted to span bases 10 through 20,inclusive</em>.</li>
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li><strong>end</strong> - The one-based ending position of the feature in the chromosome.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The end position in each BED feature is one-based. See example above</em>.</li>
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="4">
<li><strong>name</strong> - Defines the name of the BED feature.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;LINE&#8221;, &#8220;Exon3&#8221;, &#8220;HWIEAS_0001:3:1:0:266#0/1&#8221;, or &#8220;my_Feature&#8221;.</li>
<li><em>This column is optional</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="5">
<li><strong>score</strong> - The UCSC definition requires that a BED score range from 0 to 1000, inclusive. However, BEDTools allows any string to be stored in this field in order to allow greater flexibility in annotation features. For example, strings allow scientific notation for p-values, mean enrichment values, etc. It should be noted that this flexibility could prevent such annotations from being correctly displayed on the UCSC browser.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, 7.31E-05 (p-value), 0.33456 (mean enrichment value), &#8220;up&#8221;, &#8220;down&#8221;, etc.</li>
<li><em>This column is optional</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="6">
<li><strong>strand</strong> - Defines the strand - either &#8216;+&#8217; or &#8216;-&#8216;.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is optional</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="7">
<li><strong>thickStart</strong> - The starting position at which the feature is drawn thickly.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="8">
<li><strong>thickEnd</strong> - The ending position at which the feature is drawn thickly.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="9">
<li><strong>itemRgb</strong> - An RGB value of the form R,G,B (e.g. 255,0,0).</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="10">
<li><strong>blockCount</strong> - The number of blocks (exons) in the BED line.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="11">
<li><strong>blockSizes</strong> - A comma-separated list of the block sizes.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="12">
<li><strong>blockStarts</strong> - A comma-separated list of block starts.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Allowed yet ignored by BEDTools</em>.</li>
</ul>
</div></blockquote>
<p>BEDTools requires that all BED input files (and input received from stdin) are <strong>tab-delimited</strong>. The following types of BED files are supported by BEDTools:</p>
<ol class="arabic">
<li><div class="first line-block">
<div class="line"><strong>BED3</strong>: A BED file where each feature is described by <strong>chrom</strong>, <strong>start</strong>, and <strong>end</strong>.</div>
<div class="line">For example: chr1          11873   14409</div>
</div>
</li>
<li><div class="first line-block">
<div class="line"><strong>BED4</strong>: A BED file where each feature is described by <strong>chrom</strong>, <strong>start</strong>, <strong>end</strong>, and <strong>name</strong>.</div>
<div class="line">For example: chr1  11873  14409  uc001aaa.3</div>
</div>
</li>
<li><div class="first line-block">
<div class="line"><strong>BED5</strong>: A BED file where each feature is described by <strong>chrom</strong>, <strong>start</strong>, <strong>end</strong>, <strong>name</strong>, and <strong>score</strong>.</div>
<div class="line">For example: chr1 11873 14409 uc001aaa.3 0</div>
</div>
</li>
<li><div class="first line-block">
<div class="line"><strong>BED6</strong>: A BED file where each feature is described by <strong>chrom</strong>, <strong>start</strong>, <strong>end</strong>, <strong>name</strong>, <strong>score</strong>, and <strong>strand</strong>.</div>
<div class="line">For example: chr1 11873 14409 uc001aaa.3 0 +</div>
</div>
</li>
<li><div class="first line-block">
<div class="line"><strong>BED12</strong>: A BED file where each feature is described by all twelve columns listed above.</div>
<div class="line">For example: chr1 11873 14409 uc001aaa.3 0 + 11873</div>
<div class="line">11873 0 3 354,109,1189, 0,739,1347,</div>
</div>
</li>
</ol>
</div>
<div class="section" id="bedpe-format">
<h3>4.1.2 BEDPE format<a class="headerlink" href="#bedpe-format" title="Permalink to this headline">¶</a></h3>
<p>We have defined a new file format (BEDPE) in order to concisely describe disjoint genome features,
such as structural variations or paired-end sequence alignments. We chose to define a new format
because the existing &#8220;blocked&#8221; BED format (a.k.a. BED12) does not allow inter-chromosomal feature
definitions. In addition, BED12 only has one strand field, which is insufficient for paired-end sequence
alignments, especially when studying structural variation.</p>
<p>The BEDPE format is described below. The description is modified from: <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat#format1">http://genome.ucsc.edu/FAQ/FAQformat#format1</a>.</p>
<ol class="arabic simple">
<li><strong>chrom1</strong> - The name of the chromosome on which the <strong>first</strong> end of the feature exists.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;chr1&#8221;, &#8220;III&#8221;, &#8220;myChrom&#8221;, &#8220;contig1112.23&#8221;.</li>
<li><em>This column is required</em>.</li>
<li><em>Use &#8221;.&#8221; for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><strong>start1</strong> - The zero-based starting position of the <strong>first</strong> end of the feature on <strong>chrom1</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The first base in a chromosome is numbered 0</em>.</li>
<li><em>As with BED format, the start position in each BEDPE feature is therefore interpreted to be 1 greater than the start position listed in the feature. This column is required</em>.</li>
<li><em>Use -1 for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li><strong>end1</strong> - The one-based ending position of the first end of the feature on <strong>chrom1</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The end position in each BEDPE feature is one-based</em>.</li>
<li><em>This column is required</em>.</li>
<li><em>Use -1 for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="4">
<li><strong>chrom2</strong> - The name of the chromosome on which the <strong>second</strong> end of the feature exists.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;chr1&#8221;, &#8220;III&#8221;, &#8220;myChrom&#8221;, &#8220;contig1112.23&#8221;.</li>
<li><em>This column is required</em>.</li>
<li><em>Use &#8221;.&#8221; for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="5">
<li><strong>start2</strong> - The zero-based starting position of the <strong>second</strong> end of the feature on <strong>chrom2</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The first base in a chromosome is numbered 0</em>.</li>
<li><em>As with BED format, the start position in each BEDPE feature is therefore interpreted to be 1 greater than the start position listed in the feature. This column is required</em>.</li>
<li><em>Use -1 for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="6">
<li><strong>end2</strong> - The one-based ending position of the <strong>second</strong> end of the feature on <strong>chrom2</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>The end position in each BEDPE feature is one-based</em>.</li>
<li><em>This column is required</em>.</li>
<li><em>Use -1 for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="7">
<li><strong>name</strong> - Defines the name of the BEDPE feature.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;LINE&#8221;, &#8220;Exon3&#8221;, &#8220;HWIEAS_0001:3:1:0:266#0/1&#8221;, or &#8220;my_Feature&#8221;.</li>
<li><em>This column is optional</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="8">
<li><strong>score</strong> - The UCSC definition requires that a BED score range from 0 to 1000, inclusive. <em>However, BEDTools allows any string to be stored in this field in order to allow greater flexibility in annotation features</em>. For example, strings allow scientific notation for p-values, mean enrichment values, etc. It should be noted that this flexibility could prevent such annotations from being correctly displayed on the UCSC browser.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, 7.31E-05 (p-value), 0.33456 (mean enrichment value), &#8220;up&#8221;, &#8220;down&#8221;, etc.</li>
<li><em>This column is optional</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="9">
<li><strong>strand1</strong> - Defines the strand for the first end of the feature. Either &#8216;+&#8217; or &#8216;-&#8216;.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is optional</em>.</li>
<li><em>Use &#8221;.&#8221; for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="10">
<li><strong>strand2</strong> - Defines the strand for the second end of the feature. Either &#8216;+&#8217; or &#8216;-&#8216;.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is optional</em>.</li>
<li><em>Use &#8221;.&#8221; for unknown</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="11">
<li><strong>Any number of additional, user-defined fields</strong> - BEDTools allows one to add as many additional fields to the normal, 10-column BEDPE format as necessary. These columns are merely &#8220;passed through&#8221; <strong>pairToBed</strong> and <strong>pairToPair</strong> and are not part of any analysis. One would use these additional columns to add extra information (e.g., edit distance for each end of an alignment, or &#8220;deletion&#8221;, &#8220;inversion&#8221;, etc.) to each BEDPE feature.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>These additional columns are optional</em>.</li>
</ul>
</div></blockquote>
<p>Entries from an typical BEDPE file:</p>
<div class="highlight-python"><pre>chr1  100   200   chr5  5000  5100  bedpe_example1  30   +  -
chr9  1000  5000  chr9  3000  3800  bedpe_example2  100  +  -</pre>
</div>
<p>Entries from a BEDPE file with two custom fields added to each record:</p>
<div class="highlight-python"><pre>chr1  10    20    chr5  50    60    a1     30       +    -  0  1
chr9  30    40    chr9  80    90    a2     100      +    -  2  1</pre>
</div>
</div>
<div class="section" id="gff-format">
<h3>4.1.3 GFF format<a class="headerlink" href="#gff-format" title="Permalink to this headline">¶</a></h3>
<p>The GFF format is described on the Sanger Institute&#8217;s website (<a class="reference external" href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>). The GFF description below is modified from the definition at this URL. All nine columns in the GFF format description are required by BEDTools.</p>
<ol class="arabic simple">
<li><strong>seqname</strong> - The name of the sequence (e.g. chromosome) on which the feature exists.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;chr1&#8221;, &#8220;III&#8221;, &#8220;myChrom&#8221;, &#8220;contig1112.23&#8221;.</li>
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><strong>source</strong> - The source of this feature. This field will normally be used to indicate the program making the prediction, or if it comes from public database annotation, or is experimentally verified, etc.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li><strong>feature</strong> - The feature type name. Equivalent to BED&#8217;s <strong>name</strong> field.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>Any string can be used</em>. For example, &#8220;exon&#8221;, etc.</li>
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="4">
<li><strong>start</strong> - The one-based starting position of feature on <strong>seqname</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
<li><em>BEDTools accounts for the fact the GFF uses a one-based position and BED uses a zero-based start position</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="5">
<li><strong>end</strong> - The one-based ending position of feature on <strong>seqname</strong>.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="6">
<li><strong>score</strong> - A score assigned to the GFF feature. Like BED format, BEDTools allows any string to be stored in this field in order to allow greater flexibility in annotation features. We note that this differs from the GFF definition in the interest of flexibility.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="7">
<li><strong>strand</strong> - Defines the strand. Use &#8216;+&#8217;, &#8216;-&#8216; or &#8216;.&#8217;</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="8">
<li><strong>frame</strong> -  The frame of the coding sequence. Use &#8216;0&#8217;, &#8216;1&#8217;, &#8216;2&#8217;, or &#8216;.&#8217;.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="9">
<li><strong>attribute</strong> - Taken from <a class="reference external" href="http://www.sanger.ac.uk/resources/software/gff/spec.html">http://www.sanger.ac.uk/resources/software/gff/spec.html</a>: From version 2 onwards, the attribute field must have an tag value structure following the syntax used within objects in a .ace file, flattened onto one line by semicolon separators. Tags must be standard identifiers ([A-Za-z][<a href="#id2"><span class="problematic" id="id3">AZa-z0-9_</span></a>]*). Free text values must be quoted with double quotes. <em>Note: all non-printing characters in such free text value strings (e.g. newlines, tabs, control characters, etc) must be explicitly represented by their C (UNIX) style backslash-escaped representation (e.g. newlines as &#8216;n&#8217;, tabs as &#8216;t&#8217;)</em>. As in ACEDB, multiple values can follow a specific tag. The aim is to establish consistent use of particular tags, corresponding to an underlying implied ACEDB model if you want to think that way (but acedb is not required).</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><em>This column is required</em>.</li>
</ul>
</div></blockquote>
<p>An entry from an example GFF file :</p>
<div class="highlight-python"><pre>seq1 BLASTX similarity 101 235 87.1 + 0 Target "HBA_HUMAN" 11 55 ;
E_value 0.0003 dJ102G20 GD_mRNA coding_exon 7105 7201 . - 2 Sequence
"dJ102G20.C1.1"</pre>
</div>
</div>
<div class="section" id="id1">
<h3>4.1.3 GFF format<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Some of the BEDTools (e.g., genomeCoverageBed, complementBed, slopBed) need to know the size of
the chromosomes for the organism for which your BED files are based. When using the UCSC Genome
Browser, Ensemble, or Galaxy, you typically indicate which which species/genome build you are
working. The way you do this for BEDTools is to create a &#8220;genome&#8221; file, which simply lists the names of
the chromosomes (or scaffolds, etc.) and their size (in basepairs).</p>
<p>Genome files must be <strong>tab-delimited</strong> and are structured as follows (this is an example for <em>C. elegans</em>):</p>
<div class="highlight-python"><pre>chrI  15072421
chrII 15279323
...
chrX  17718854
chrM  13794</pre>
</div>
<p>BEDTools includes pre-defined genome files for human and mouse in the <strong>/genomes</strong> directory included
in the BEDTools distribution.</p>
</div>
<div class="section" id="sam-bam-format">
<h3>4.1.5 SAM/BAM format<a class="headerlink" href="#sam-bam-format" title="Permalink to this headline">¶</a></h3>
<p>The SAM / BAM format is a powerful and widely-used format for storing sequence alignment data (see
<a class="reference external" href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a> for more details). It has quickly become the standard format to which
most DNA sequence alignment programs write their output. Currently, the following BEDTools
support inout in BAM format: <em>intersectBed, windowBed, coverageBed, genomeCoverageBed,
pairToBed, bamToBed</em>. Support for the BAM format in BEDTools allows one to (to name a few):
compare sequence alignments to annotations, refine alignment datasets, screen for potential mutations
and compute aligned sequence coverage.</p>
<p>The details of how these tools work with BAM files are addressed in <strong>Section 5</strong> of this manual.</p>
</div>
<div class="section" id="vcf-format">
<h3>4.1.6 VCF format<a class="headerlink" href="#vcf-format" title="Permalink to this headline">¶</a></h3>
<p>The Variant Call Format (VCF) was conceived as part of the 1000 Genomes Project as a standardized
means to report genetic variation calls from SNP, INDEL and structural variant detection programs
(see <a class="reference external" href="http://www.1000genomes.org/wiki/doku.php?id=1000_genomes:analysis:vcf4.0">http://www.1000genomes.org/wiki/doku.php?id=1000_genomes:analysis:vcf4.0</a> for details).
BEDTools now supports the latest version of this format (i.e, Version 4.0). As a result, BEDTools can
be used to compare genetic variation calls with other genomic features.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/bedtools.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">General usage</a><ul>
<li><a class="reference internal" href="#supported-file-formats">4.1 Supported file formats</a><ul>
<li><a class="reference internal" href="#bed-format">4.1.1 BED format</a></li>
<li><a class="reference internal" href="#bedpe-format">4.1.2 BEDPE format</a></li>
<li><a class="reference internal" href="#gff-format">4.1.3 GFF format</a></li>
<li><a class="reference internal" href="#id1">4.1.3 GFF format</a></li>
<li><a class="reference internal" href="#sam-bam-format">4.1.5 SAM/BAM format</a></li>
<li><a class="reference internal" href="#vcf-format">4.1.6 VCF format</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="quick-start.html"
                        title="previous chapter">Quick start</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bedtools-suite.html"
                        title="next chapter">The BEDTools suite</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/general-usage.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="bedtools-suite.html" title="The BEDTools suite"
             >next</a> |</li>
        <li class="right" >
          <a href="quick-start.html" title="Quick start"
             >previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012.
      Last updated on Jul 23, 2012.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>
  </body>
</html>