

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Overview</title>
    
    <link rel="stylesheet" href="../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.16.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/bedtools.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="bedtools: a powerful toolset for genome arithmetic" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="bedtools: a powerful toolset for genome arithmetic"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<div class="section" id="background">
<h2>1.1 Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h2>
<p>The development of BEDTools was motivated by a need for fast, flexible tools with which to compare large sets of genomic
features. Answering fundamental research questions with existing tools was either too slow or required modifications to the
way they reported or computed their results. We were aware of the utilities on the UCSC Genome Browser and Galaxy websites, as
well as the elegant tools available as part of Jim Kent’s monolithic suite of tools (“Kent source”). However, we found that
the web-based tools were too cumbersome when working with large datasets generated by current sequencing technologies.
Similarly, we found that the Kent source command line tools often required a local installation of the UCSC Genome Browser.
These limitations, combined with the fact that we often wanted an extra option here or there that wasn’t available with
existing tools, led us to develop our own from scratch. The initial version of BEDTools was publicly released in the spring of
2009. The current version has evolved from our research experiences and those of the scientists using the suite over the last
year. The BEDTools suite enables one to answer common questions of genomic data in a fast and reliable manner. The fact that
almost all the utilities accept input from “stdin” allows one to “stream / pipe” several commands together to facilitate more
complicated analyses. Also, the tools allow fine control over how output is reported. The initial version of BEDTools
supported solely 6-column <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat#format1">BED</a> files. <em>However, we have subsequently added support for sequence alignments in</em> <a class="reference external" href="http://samtools.sourceforge.net/">BAM</a>
<em>format, as well as for features in</em> <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat#format3">GFF</a> , <em>“blocked” BED format, and</em>
<a class="reference external" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a> <em>format</em>.
The tools are quite fast and typically finish in a matter of a few seconds, even for large datasets. This manual seeks to describe the behavior and
available functionality for each BEDTool. Usage examples are scattered throughout the text, and formal examples are
provided in the last two sections, we hope that this document will give you a sense of the flexibility of
the toolkit and the types of analyses that are possible with BEDTools. If you have further questions, please join the BEDTools
discussion group, visit the Usage Examples on the Google Code site (usage, advanced usage), or take a look at the nascent
“Usage From the Wild” page.</p>
</div>
<div class="section" id="summary-of-available-tools">
<h2>1.2 Summary of available tools.<a class="headerlink" href="#summary-of-available-tools" title="Permalink to this headline">¶</a></h2>
<p>BEDTools support a  wide range of operations for  interrogating and manipulating genomic features. The table below summarizes
the tools available in the suite.</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Utility</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><strong>intersectBed</strong></td>
<td>Returns overlaps between two BED/GFF/VCF files.</td>
</tr>
<tr class="row-odd"><td><strong>pairToBed</strong></td>
<td>Returns overlaps between a paired-end BED file and a regular BED/VCF/GFF file.</td>
</tr>
<tr class="row-even"><td><strong>bamToBed</strong></td>
<td>Converts BAM alignments to BED6, BED12, or BEDPE format.</td>
</tr>
<tr class="row-odd"><td><strong>bedToBam</strong></td>
<td>Converts BED/GFF/VCF features to BAM format.</td>
</tr>
<tr class="row-even"><td><strong>bed12ToBed6</strong></td>
<td>Converts &#8220;blocked&#8221; BED12 features to discrete BED6 features.</td>
</tr>
<tr class="row-odd"><td><strong>bedToIgv</strong></td>
<td>Creates IGV batch scripts for taking multiple snapshots from BED/GFF/VCF features.</td>
</tr>
<tr class="row-even"><td><strong>coverageBed</strong></td>
<td>Summarizes the depth and breadth of coverage of features in one BED versus features (e.g, windows, exons, etc.) defined in another BED/GFF/VCF file.</td>
</tr>
<tr class="row-odd"><td><strong>multiBamCov</strong></td>
<td>Counts sequence coverage for multiple position-sorted bams at specific loci defined in a BED/GFF/VCF file</td>
</tr>
<tr class="row-even"><td><strong>tagBam</strong></td>
<td>Annotates a BAM file with custom tag fields based on overlaps with BED/GFF/VCF files</td>
</tr>
<tr class="row-odd"><td><strong>nuclBed</strong></td>
<td>Profiles the nucleotide content of intervals in a fasta file</td>
</tr>
<tr class="row-even"><td><strong>genomeCoverageBed</strong></td>
<td>Creates either a histogram, BEDGRAPH, or a &#8220;per base&#8221; report of genome coverage.</td>
</tr>
<tr class="row-odd"><td><strong>unionBedGraphs</strong></td>
<td>Combines multiple BedGraph? files into a single file, allowing coverage/other comparisons between them.</td>
</tr>
<tr class="row-even"><td><strong>annotateBed</strong></td>
<td>Annotates one BED/VCF/GFF file with overlaps from many others.</td>
</tr>
<tr class="row-odd"><td><strong>groupBy</strong></td>
<td>Deprecated. Now in the filo package.</td>
</tr>
<tr class="row-even"><td><strong>overlap</strong></td>
<td>Returns the number of bases pairs of overlap b/w two features on the same line.</td>
</tr>
<tr class="row-odd"><td><strong>pairToPair</strong></td>
<td>Returns overlaps between two paired-end BED files.</td>
</tr>
<tr class="row-even"><td><strong>closestBed</strong></td>
<td>Returns the closest feature to each entry in a BED/GFF/VCF file.</td>
</tr>
<tr class="row-odd"><td><strong>subtractBed</strong></td>
<td>Removes the portion of an interval that is overlapped by another feature.</td>
</tr>
<tr class="row-even"><td><strong>windowBed</strong></td>
<td>Returns overlaps between two BED/VCF/GFF files based on a user-defined window.</td>
</tr>
<tr class="row-odd"><td><strong>mergeBed</strong></td>
<td>Merges overlapping features into a single feature.</td>
</tr>
<tr class="row-even"><td><strong>complementBed</strong></td>
<td>Returns all intervals not spanned by the features in a BED/GFF/VCF file.</td>
</tr>
<tr class="row-odd"><td><strong>fastaFromBed</strong></td>
<td>Creates FASTA sequences based on intervals in a BED/GFF/VCF file.</td>
</tr>
<tr class="row-even"><td><strong>maskFastaFromBed</strong></td>
<td>Masks a FASTA file based on BED coordinates.</td>
</tr>
<tr class="row-odd"><td><strong>shuffleBed</strong></td>
<td>Randomly permutes the locations of a BED file among a genome.</td>
</tr>
<tr class="row-even"><td><strong>slopBed</strong></td>
<td>Adjusts each BED entry by a requested number of base pairs.</td>
</tr>
<tr class="row-odd"><td><strong>flankBed</strong></td>
<td>Creates flanking intervals for each feature in a BED/GFF/VCF file.</td>
</tr>
<tr class="row-even"><td><strong>sortBed</strong></td>
<td>Sorts a BED file by chrom, then start position. Other ways as well.</td>
</tr>
<tr class="row-odd"><td><strong>linksBed</strong></td>
<td>Creates an HTML file of links to the UCSC or a custom browser.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="fundamental-concepts">
<h2>1.3 Fundamental concepts.<a class="headerlink" href="#fundamental-concepts" title="Permalink to this headline">¶</a></h2>
<div class="section" id="what-are-genome-features-and-how-are-they-represented">
<h3>1.3.1 What are genome features and how are they represented?<a class="headerlink" href="#what-are-genome-features-and-how-are-they-represented" title="Permalink to this headline">¶</a></h3>
<p>Throughout this manual, we will discuss how to use BEDTools to manipulate, compare and ask questions of genome “features”. Genome features can be functional elements (e.g., genes), genetic polymorphisms (e.g.
SNPs, INDELs, or structural variants), or other annotations that have been discovered or curated by genome sequencing groups or genome browser groups. In addition, genome features can be custom annotations that
an individual lab or researcher defines (e.g., my novel gene or variant).</p>
<p>The basic characteristics of a genome feature are the chromosome or scaffold on which the feature “resides”, the base pair on which the
feature starts (i.e. the “start”), the base pair on which feature ends (i.e. the “end”), the strand on which the feature exists (i.e. “+” or “-“), and the name of the feature if one is applicable.</p>
<p>The two most widely used formats for representing genome features are the BED (Browser Extensible Data) and GFF (General Feature Format) formats. BEDTools was originally written to work exclusively with genome features
described using the BED format, but it has been recently extended to seamlessly work with BED, GFF and VCF files.</p>
<p>Existing annotations for the genomes of many species can be easily downloaded in BED and GFF
format from the UCSC Genome Browser’s “Table Browser” (<a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables?command=start">http://genome.ucsc.edu/cgi-bin/hgTables?command=start</a>) or from the “Bulk Downloads” page (<a class="reference external" href="http://hgdownload.cse.ucsc.edu/downloads.html">http://hgdownload.cse.ucsc.edu/downloads.html</a>). In addition, the
Ensemble Genome Browser contains annotations in GFF/GTF format for many species (<a class="reference external" href="http://www.ensembl.org/info/data/ftp/index.html">http://www.ensembl.org/info/data/ftp/index.html</a>)</p>
</div>
<div class="section" id="overlapping-intersecting-features">
<h3>1.3.2 Overlapping / intersecting features.<a class="headerlink" href="#overlapping-intersecting-features" title="Permalink to this headline">¶</a></h3>
<p>Two genome features (henceforth referred to as “features”) are said to overlap or intersect if they share at least one base in common.
In the figure below, Feature A intersects/overlaps Feature B, but it does not intersect/overlap Feature C.</p>
<p><strong>TODO: place figure here</strong></p>
</div>
<div class="section" id="comparing-features-in-file-a-and-file-b">
<h3>1.3.3 Comparing features in file “A” and file “B”.<a class="headerlink" href="#comparing-features-in-file-a-and-file-b" title="Permalink to this headline">¶</a></h3>
<p>The previous section briefly introduced a fundamental naming convention used in BEDTools. Specifically, all BEDTools that compare features contained in two distinct files refer to one file as feature set “A” and the other file as feature set “B”. This is mainly in the interest of brevity, but it also has its roots in set theory.
As an example, if one wanted to look for SNPs (file A) that overlap with exons (file B), one would use intersectBed in the following manner:</p>
<div class="highlight-python"><pre>intersectBed –a snps.bed –b exons.bed</pre>
</div>
<p>There are two exceptions to this rule: 1) When the “A” file is in BAM format, the “-abam” option must bed used. For example:</p>
<div class="highlight-python"><pre>intersectBed –abam alignedReads.bam –b exons.bed</pre>
</div>
<p>And 2) For tools where only one input feature file is needed, the “-i” option is used. For example:</p>
<div class="highlight-python"><pre>mergeBed –i repeats.bed</pre>
</div>
</div>
<div class="section" id="bed-starts-are-zero-based-and-bed-ends-are-one-based">
<h3>1.3.4 BED starts are zero-based and BED ends are one-based.<a class="headerlink" href="#bed-starts-are-zero-based-and-bed-ends-are-one-based" title="Permalink to this headline">¶</a></h3>
<p>BEDTools users are sometimes confused by the way the start and end of BED features are represented. Specifically, BEDTools uses the UCSC Genome Browser’s internal database convention of making the start position 0-based and the end position 1-based: (<a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQtracks#tracks1">http://genome.ucsc.edu/FAQ/FAQtracks#tracks1</a>)
In other words, BEDTools interprets the “start” column as being 1 basepair higher than what is represented in the file. For example, the following BED feature represents a single base on chromosome 1; namely, the 1st base:</p>
<div class="highlight-python"><pre>chr1   0        1    first_base</pre>
</div>
<p>Why, you might ask? The advantage of storing features this way is that when computing the length of a feature, one must simply subtract the start from the end. Were the start position 1-based,
the calculation would be (slightly) more complex (i.e. (end-start)+1). Thus, storing BED features this way reduces the computational burden.</p>
</div>
<div class="section" id="gff-starts-and-ends-are-one-based">
<h3>1.3.5 GFF starts and ends are one-based.<a class="headerlink" href="#gff-starts-and-ends-are-one-based" title="Permalink to this headline">¶</a></h3>
<p>In contrast, the GFF format uses 1-based coordinates for both the start and the end positions. BEDTools is aware of this and adjusts the positions accordingly.
In other words, you don’t need to subtract 1 from the start positions of your GFF features for them to work correctly with BEDTools.</p>
</div>
<div class="section" id="vcf-coordinates-are-one-based">
<h3>1.3.6 VCF coordinates are one-based.<a class="headerlink" href="#vcf-coordinates-are-one-based" title="Permalink to this headline">¶</a></h3>
<p>The VCF format uses 1-based coordinates. As in GFF, BEDTools is aware of this and adjusts the positions accordingly.
In other words, you don’t need to subtract 1 from the start positions of your VCF features for them to work correctly with BEDTools.</p>
</div>
<div class="section" id="file-b-is-loaded-into-memory-most-of-the-time">
<h3>1.3.7 File B is loaded into memory (most of the time).<a class="headerlink" href="#file-b-is-loaded-into-memory-most-of-the-time" title="Permalink to this headline">¶</a></h3>
<p>Whenever a BEDTool compares two files of features, the “B” file is loaded into memory. By contrast, the “A” file is processed line by line and compared with the features from B.
Therefore to minimize memory usage, one should set the smaller of the two files as the B file. One salient example is the comparison of aligned sequence reads from a
current DNA sequencer to gene annotations.      In this case, the aligned sequence file (in BED format) may have tens of millions of features (the sequence alignments),
while the gene annotation file will have tens of thousands of features. In this case, it is wise to sets the reads as file A and the genes as file B.</p>
</div>
<div class="section" id="feature-files-must-be-tab-delimited">
<h3>1.3.8 Feature files <em>must</em> be tab-delimited.<a class="headerlink" href="#feature-files-must-be-tab-delimited" title="Permalink to this headline">¶</a></h3>
<p>This is rather self-explanatory. While it is possible to allow BED files to be space-delimited, we have decided to require tab delimiters for three reasons:</p>
<ol class="arabic simple">
<li>By requiring one delimiter type, the processing time is minimized.</li>
<li>Tab-delimited files are more amenable to other UNIX utilities.</li>
<li>GFF files can contain spaces within attribute columns. This complicates the use of space-delimited files as spaces must therefore be treated specially depending on the context.</li>
</ol>
</div>
<div class="section" id="all-bedtools-allow-features-to-be-piped-via-standard-input">
<h3>1.3.9 All BEDTools allow features to be “piped” via standard input.<a class="headerlink" href="#all-bedtools-allow-features-to-be-piped-via-standard-input" title="Permalink to this headline">¶</a></h3>
<p>In an effort to allow one to combine multiple BEDTools and other UNIX utilities into more complicated “pipelines”, all BEDTools allow features
to be passed to them via standard input. Only one feature file may be passed to a BEDTool via standard input.
The convention used by all BEDTools is to set either file A or file B to “stdin” or &#8220;-&#8221;. For example:</p>
<div class="highlight-python"><pre>cat snps.bed | intersectBed –a stdin –b exons.bed
cat snps.bed | intersectBed –a - –b exons.bed</pre>
</div>
<p>In addition, all BEDTools that simply require one main input file (the -i file) will assume that input is
coming from standard input if the -i parameter is ignored. For example, the following are equivalent:</p>
<div class="highlight-python"><pre>cat snps.bed | sortBed –i stdin
cat snps.bed | sortBed</pre>
</div>
</div>
<div class="section" id="most-bedtools-write-their-results-to-standard-output">
<h3>1.3.10 Most BEDTools write their results to standard output.<a class="headerlink" href="#most-bedtools-write-their-results-to-standard-output" title="Permalink to this headline">¶</a></h3>
<p>To allow one to combine multiple BEDTools and other UNIX utilities into more complicated “pipelines”,
most BEDTools report their output to standard output, rather than to a named file. If one wants to write the output to a named file, one can use the UNIX “file redirection” symbol “&gt;” to do so.
Writing to standard output (the default):</p>
<div class="highlight-python"><pre>intersectBed –a snps.bed –b exons.bed
chr1 100100 100101 rs233454
chr1 200100 200101 rs446788
chr1 300100 300101 rs645678</pre>
</div>
<p>Writing to a file:</p>
<div class="highlight-python"><pre>intersectBed –a snps.bed –b exons.bed &gt; snps.in.exons.bed

cat snps.in.exons.bed
chr1 100100 100101 rs233454
chr1 200100 200101 rs446788
chr1 300100 300101 rs645678</pre>
</div>
</div>
<div class="section" id="what-is-a-genome-file">
<h3>1.3.11 What is a “genome” file?<a class="headerlink" href="#what-is-a-genome-file" title="Permalink to this headline">¶</a></h3>
<p>Some of the BEDTools (e.g., genomeCoverageBed, complementBed, slopBed) need to know the size of
the chromosomes for the organism for which your BED files are based. When using the UCSC Genome
Browser, Ensemble, or Galaxy, you typically indicate which species / genome build you are working.
The way you do this for BEDTools is to create a “genome” file, which simply lists the names of the
chromosomes (or scaffolds, etc.) and their size (in basepairs).
Genome files must be tab-delimited and are structured as follows (this is an example for C. elegans):</p>
<div class="highlight-python"><pre>chrI 15072421
chrII 15279323
...
chrX 17718854
chrM 13794</pre>
</div>
<p>BEDTools includes predefined genome files for human and mouse in the /genomes directory included
in the BEDTools distribution. Additionally, the “chromInfo” files/tables available from the UCSC
Genome Browser website are acceptable. For example, one can download the hg19 chromInfo file here:
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz</a></p>
</div>
<div class="section" id="paired-end-bed-files-bedpe-files">
<h3>1.3.12 Paired-end BED files (BEDPE files).<a class="headerlink" href="#paired-end-bed-files-bedpe-files" title="Permalink to this headline">¶</a></h3>
<p>We have defined a new file format (BEDPE) to concisely describe disjoint genome features, such as
structural variations or paired-end sequence alignments. We chose to define a new format because the
existing BED block format (i.e. BED12) does not allow inter-chromosomal feature definitions. Moreover,
the BED12 format feels rather bloated when one want to describe events with only two blocks.</p>
</div>
<div class="section" id="use-h-for-help-with-any-bedtool">
<h3>1.3.13 Use “-h” for help with any BEDTool.<a class="headerlink" href="#use-h-for-help-with-any-bedtool" title="Permalink to this headline">¶</a></h3>
<p>Rather straightforward. If you use the “-h” option with any BEDTool, a full menu of example usage
and available options (when applicable) will be reported.</p>
</div>
<div class="section" id="bed-features-must-not-contain-negative-positions">
<h3>1.3.14 BED features must not contain negative positions.<a class="headerlink" href="#bed-features-must-not-contain-negative-positions" title="Permalink to this headline">¶</a></h3>
<p>BEDTools will typically reject BED features that contain negative positions. In special cases, however,
BEDPE positions may be set to -1 to indicate that one or more ends of a BEDPE feature is unaligned.</p>
</div>
<div class="section" id="the-start-position-must-be-to-the-end-position">
<h3>1.3.15 The start position must be &lt;= to the end position.<a class="headerlink" href="#the-start-position-must-be-to-the-end-position" title="Permalink to this headline">¶</a></h3>
<p>BEDTools will reject BED features where the start position is greater than the end position.</p>
</div>
<div class="section" id="headers-are-allowed-in-gff-and-bed-files">
<h3>1.3.16 Headers are allowed in GFF and BED files<a class="headerlink" href="#headers-are-allowed-in-gff-and-bed-files" title="Permalink to this headline">¶</a></h3>
<p>BEDTools will ignore headers at the beginning of BED and GFF files. Valid header lines begin with a
“#” symbol, the work “track”, or the word “browser”. For example, the following examples are valid
headers for BED or GFF files:</p>
<div class="highlight-python"><pre>track name=aligned_read description="Illumina aligned reads”
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 -

#This is a fascinating dataset
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 -

browser position chr22:1-20000
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 -</pre>
</div>
</div>
<div class="section" id="gzip-support-bed-gff-vcf-and-bedpe-file-can-be-gzipped">
<h3>1.3.17 GZIP support: BED, GFF, VCF, and BEDPE file can be “gzipped”<a class="headerlink" href="#gzip-support-bed-gff-vcf-and-bedpe-file-can-be-gzipped" title="Permalink to this headline">¶</a></h3>
<p>BEDTools will process gzipped BED, GFF, VCF and BEDPE files in the same manner as
uncompressed files. Gzipped files are auto-detected thanks to a helpful contribution from Gordon
Assaf.</p>
</div>
<div class="section" id="support-for-split-or-spliced-bam-alignments-and-blocked-bed-features">
<h3>1.3.18 Support for “split” or “spliced” BAM alignments and “blocked” BED features<a class="headerlink" href="#support-for-split-or-spliced-bam-alignments-and-blocked-bed-features" title="Permalink to this headline">¶</a></h3>
<p>As of Version 2.8.0, five BEDTools (<tt class="docutils literal"><span class="pre">intersectBed</span></tt>, <tt class="docutils literal"><span class="pre">coverageBed</span></tt>, <tt class="docutils literal"><span class="pre">genomeCoverageBed</span></tt>,
<tt class="docutils literal"><span class="pre">bamToBed</span></tt>, and <tt class="docutils literal"><span class="pre">bed12ToBed6</span></tt>) can properly handle “split”/”spliced” BAM alignments (i.e., having an
“N” CIGAR operation) and/or “blocked” BED (aka BED12) features.</p>
<p><tt class="docutils literal"><span class="pre">intersectBed</span></tt>, <tt class="docutils literal"><span class="pre">coverageBed</span></tt>, and <tt class="docutils literal"><span class="pre">genomeCoverageBed</span></tt> will optionally handle “split” BAM and/or
“blocked” BED by using the <tt class="docutils literal"><span class="pre">-split</span></tt> option. This will cause intersects or coverage to be computed only
for the alignment or feature blocks. In contrast, without this option, the intersects/coverage would be
computed for the entire “span” of the alignment or feature, regardless of the size of the gaps between
each alignment or feature block. For example, imagine you have a RNA-seq read that originates from
the junction of two exons that were spliced together in a mRNA. In the genome, these two exons
happen to be 30Kb apart. Thus, when the read is aligned to the reference genome, one portion of the
read will align to the first exon, while another portion of the read will align ca. 30Kb downstream to the
other exon. The corresponding CIGAR string would be something like (assuming a 76bp read):
30M*3000N*46M. In the genome, this alignment “spans” 3076 bp, yet the nucleotides in the sequencing
read only align “cover” 76bp. Without the <tt class="docutils literal"><span class="pre">-split</span></tt> option, coverage or overlaps would be reported for the
entire 3076bp span of the alignment. However, with the <tt class="docutils literal"><span class="pre">-split</span></tt> option, coverage or overlaps will only
be reported for the portions of the read that overlap the exons (i.e. 30bp on one exon, and
46bp on the other).</p>
<p>Using the -split option with bamToBed causes “spliced/split” alignments to be reported in BED12
format. Using the -split option with bed12ToBed6 causes “blocked” BED12 features to be reported in
BED6 format.</p>
</div>
<div class="section" id="writing-uncompressed-bam-output">
<h3>1.3.19 Writing uncompressed BAM output.<a class="headerlink" href="#writing-uncompressed-bam-output" title="Permalink to this headline">¶</a></h3>
<p>When working with a large BAM file using a complex set of tools in a pipe/stream, it is advantageous
to pass uncompressed BAM output to each downstream program. This minimizes the amount of time
spent compressing and decompressing output from one program to the next. All BEDTools that create
BAM output (e.g. <tt class="docutils literal"><span class="pre">intersectBed</span></tt>, <tt class="docutils literal"><span class="pre">windowBed</span></tt>) will now optionally create uncompressed BAM output
using the <tt class="docutils literal"><span class="pre">-ubam</span></tt> option.</p>
</div>
</div>
<div class="section" id="implementation-and-algorithmic-notes">
<h2>1.4 Implementation and algorithmic notes.<a class="headerlink" href="#implementation-and-algorithmic-notes" title="Permalink to this headline">¶</a></h2>
<p>BEDTools was implemented in C++ and makes extensive use of data structures and fundamental
algorithms from the Standard Template Library (STL). Many of the core algorithms are based upon the
genome binning algorithm described in the original UCSC Genome Browser paper (Kent et al, 2002).
The tools have been designed to inherit core data structures from central source files, thus allowing
rapid tool development and deployment of improvements and corrections. Support for BAM files is
made possible through Derek Barnett’s elegant C++ API called BamTools.</p>
</div>
<div class="section" id="license-and-availability">
<h2>1.5 License and availability.<a class="headerlink" href="#license-and-availability" title="Permalink to this headline">¶</a></h2>
<p>BEDTools is freely available under a GNU Public License (Version 2) at:
<a class="reference external" href="http://bedtools.googlecode.com">http://bedtools.googlecode.com</a></p>
</div>
<div class="section" id="mailing-list">
<h2>1.6 Mailing list.<a class="headerlink" href="#mailing-list" title="Permalink to this headline">¶</a></h2>
<p>A discussion group for reporting bugs, asking questions of the developer and of the user community, as
well as for requesting new features is available at:
<a class="reference external" href="http://groups.google.com/group/bedtools-discuss">http://groups.google.com/group/bedtools-discuss</a></p>
</div>
<div class="section" id="contributors">
<h2>1.7 Contributors.<a class="headerlink" href="#contributors" title="Permalink to this headline">¶</a></h2>
<p>As open-source software, BEDTools greatly benefits from contributions made by other developers and
users of the tools. We encourage and welcome suggestions, contributions and complaints. This is how
software matures, improves and stays on top of the needs of its user community. The Google Code
(GC) site maintains a list of individuals who have contributed either source code or useful ideas for
improving the tools. In the near future, we hope to maintain a source repository on the GC site in
order to facilitate further contributions. We are currently unable to do so because we use Git for
version control, which is not yet supported by GC.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/bedtools.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Overview</a><ul>
<li><a class="reference internal" href="#background">1.1 Background</a></li>
<li><a class="reference internal" href="#summary-of-available-tools">1.2 Summary of available tools.</a></li>
<li><a class="reference internal" href="#fundamental-concepts">1.3 Fundamental concepts.</a><ul>
<li><a class="reference internal" href="#what-are-genome-features-and-how-are-they-represented">1.3.1 What are genome features and how are they represented?</a></li>
<li><a class="reference internal" href="#overlapping-intersecting-features">1.3.2 Overlapping / intersecting features.</a></li>
<li><a class="reference internal" href="#comparing-features-in-file-a-and-file-b">1.3.3 Comparing features in file “A” and file “B”.</a></li>
<li><a class="reference internal" href="#bed-starts-are-zero-based-and-bed-ends-are-one-based">1.3.4 BED starts are zero-based and BED ends are one-based.</a></li>
<li><a class="reference internal" href="#gff-starts-and-ends-are-one-based">1.3.5 GFF starts and ends are one-based.</a></li>
<li><a class="reference internal" href="#vcf-coordinates-are-one-based">1.3.6 VCF coordinates are one-based.</a></li>
<li><a class="reference internal" href="#file-b-is-loaded-into-memory-most-of-the-time">1.3.7 File B is loaded into memory (most of the time).</a></li>
<li><a class="reference internal" href="#feature-files-must-be-tab-delimited">1.3.8 Feature files <em>must</em> be tab-delimited.</a></li>
<li><a class="reference internal" href="#all-bedtools-allow-features-to-be-piped-via-standard-input">1.3.9 All BEDTools allow features to be “piped” via standard input.</a></li>
<li><a class="reference internal" href="#most-bedtools-write-their-results-to-standard-output">1.3.10 Most BEDTools write their results to standard output.</a></li>
<li><a class="reference internal" href="#what-is-a-genome-file">1.3.11 What is a “genome” file?</a></li>
<li><a class="reference internal" href="#paired-end-bed-files-bedpe-files">1.3.12 Paired-end BED files (BEDPE files).</a></li>
<li><a class="reference internal" href="#use-h-for-help-with-any-bedtool">1.3.13 Use “-h” for help with any BEDTool.</a></li>
<li><a class="reference internal" href="#bed-features-must-not-contain-negative-positions">1.3.14 BED features must not contain negative positions.</a></li>
<li><a class="reference internal" href="#the-start-position-must-be-to-the-end-position">1.3.15 The start position must be &lt;= to the end position.</a></li>
<li><a class="reference internal" href="#headers-are-allowed-in-gff-and-bed-files">1.3.16 Headers are allowed in GFF and BED files</a></li>
<li><a class="reference internal" href="#gzip-support-bed-gff-vcf-and-bedpe-file-can-be-gzipped">1.3.17 GZIP support: BED, GFF, VCF, and BEDPE file can be “gzipped”</a></li>
<li><a class="reference internal" href="#support-for-split-or-spliced-bam-alignments-and-blocked-bed-features">1.3.18 Support for “split” or “spliced” BAM alignments and “blocked” BED features</a></li>
<li><a class="reference internal" href="#writing-uncompressed-bam-output">1.3.19 Writing uncompressed BAM output.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#implementation-and-algorithmic-notes">1.4 Implementation and algorithmic notes.</a></li>
<li><a class="reference internal" href="#license-and-availability">1.5 License and availability.</a></li>
<li><a class="reference internal" href="#mailing-list">1.6 Mailing list.</a></li>
<li><a class="reference internal" href="#contributors">1.7 Contributors.</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter"><strong>bedtools</strong>: <em>a powerful toolset for genome arithmetic</em></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installation</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/overview.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="bedtools: a powerful toolset for genome arithmetic"
             >previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012.
      Last updated on Jul 23, 2012.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>
  </body>
</html>