

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>5.2 pairToBed</title>
    
    <link rel="stylesheet" href="../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.16.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/bedtools.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="up" title="The BEDTools suite" href="bedtools-suite.html" />
    <link rel="next" title="5.3 pairToPair" href="pairToPair.html" />
    <link rel="prev" title="5.1 intersect" href="intersectBed.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="pairToPair.html" title="5.3 pairToPair"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="intersectBed.html" title="5.1 intersect"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li>
          <li><a href="bedtools-suite.html" accesskey="U">The BEDTools suite</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pairtobed">
<h1>5.2 pairToBed<a class="headerlink" href="#pairtobed" title="Permalink to this headline">¶</a></h1>
<p><strong>pairToBed</strong> compares each end of a BEDPE feature or a paired-end BAM alignment to a feature file in
search of overlaps.</p>
<p><strong>NOTE: pairToBed requires that the BAM file is sorted/grouped by the read name. This
allows pairToBed to extract correct alignment coordinates for each end based on their
respective CIGAR strings. It also assumes that the alignments for a given pair come in
groups of twos. There is not yet a standard method for reporting multiple alignments
using BAM. pairToBed will fail if an aligner does not report alignments in pairs.</strong></p>
<div class="section" id="usage-and-option-summary">
<h2>5.2.1 Usage and option summary<a class="headerlink" href="#usage-and-option-summary" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage:</strong></p>
<div class="highlight-python"><pre>pairToBed [OPTIONS] [-a &lt;BEDPE&gt; || -abam &lt;BAM&gt;] -b &lt;BED/GFF/VCF&gt;</pre>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><strong>-a</strong></td>
<td>BEDPE file A. Each feature in A is compared to B in search of overlaps. Use &#8220;stdin&#8221; if passing A with a UNIX pipe. Output will be in BEDPE format.</td>
</tr>
<tr class="row-odd"><td><strong>-b</strong></td>
<td>BED file B. Use &#8220;stdin&#8221; if passing B with a UNIX pipe.</td>
</tr>
<tr class="row-even"><td><strong>-abam</strong></td>
<td>BAM file A. Each end of each BAM alignment in A is compared to B in search of overlaps. Use &#8220;stdin&#8221; if passing A with a UNIX pipe: For example: samtools view ?Cb &lt;BAM&gt; | pairToBed ?Cabam stdin ?Cb genes.bed | samtools view -</td>
</tr>
<tr class="row-odd"><td><strong>-ubam</strong></td>
<td>Write uncompressed BAM output. The default is write compressed BAM output.</td>
</tr>
<tr class="row-even"><td><strong>-bedpe</strong></td>
<td>When using BAM input (-abam), write output as BEDPE. The default is to write output in BAM when using -abam. For example: pairToBed ?Cabam reads.bam ?Cb genes.bed ?Cbedpe</td>
</tr>
<tr class="row-odd"><td><strong>-ed</strong></td>
<td>Use BAM total edit distance (NM tag) for BEDPE score. Default for BEDPE is to use the <em>minimum</em> of the two mapping qualities for the pair. When -ed is used the <em>total</em> edit distance from the two mates is reported as the score.</td>
</tr>
<tr class="row-even"><td><strong>-f</strong></td>
<td>Minimum overlap required as a fraction of A. Default is 1E-9 (i.e. 1bp).</td>
</tr>
<tr class="row-odd"><td><strong>-s</strong></td>
<td>Force &#8220;strandedness&#8221;. That is, only report hits in B that overlap A on the <strong>same</strong> strand. By default, overlaps are reported without respect to strand.</td>
</tr>
<tr class="row-even"><td><strong>-type</strong></td>
<td><p class="first">Approach to reporting overlaps between BEDPE and BED.</p>
<blockquote class="last">
<div><blockquote>
<div><p><strong>either-</strong> Report overlaps if either end of A overlaps B.</p>
<ul class="simple">
<li><em>Default</em></li>
</ul>
<p><strong>neither-</strong> Report A if neither end of A overlaps B.</p>
<p><strong>xor-</strong> Report overlaps if one and only one end of A overlaps B.</p>
<p><strong>both-</strong> Report overlaps if both ends of A overlap B.</p>
<p><strong>notboth-</strong> Report overlaps if neither end or one and only one end of A overlap B.</p>
<p><strong>ispan-</strong> Report overlaps between [end1, start2] of A and B.</p>
<ul class="simple">
<li>Note: If chrom1 &lt;&gt; chrom2, entry is ignored.</li>
</ul>
</div></blockquote>
<p><strong>ospan-</strong> Report overlaps between [start1, end2] of A and B.</p>
<blockquote>
<div><ul class="simple">
<li>Note: If chrom1 &lt;&gt; chrom2, entry is ignored.</li>
</ul>
<p><strong>notispan-</strong>  Report A if ispan of A doesn&#8217;t overlap B.
- Note: If chrom1 &lt;&gt; chrom2, entry is ignored.</p>
<p><strong>notospan-</strong>  Report A if ospan of A doesn&#8217;t overlap B.
- Note: If chrom1 &lt;&gt; chrom2, entry is ignored.</p>
</div></blockquote>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="default-behavior">
<h2>5.2.2 Default behavior<a class="headerlink" href="#default-behavior" title="Permalink to this headline">¶</a></h2>
<p>By default, a BEDPE / BAM feature will be reported if <em>either</em> end overlaps a feature in the BED file.
In the example below, the left end of the pair overlaps B yet the right end does not. Thus, BEDPE/
BAM A is reported since the default is to report A if either end overlaps B.</p>
<p>Default: Report A if <em>either</em> end overlaps B.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                          ^^^^^^

Result              =====.................................=====</pre>
</div>
</div>
<div class="section" id="type-optional-overlap-requirements">
<h2>5.2.3 (-type)Optional overlap requirements<a class="headerlink" href="#type-optional-overlap-requirements" title="Permalink to this headline">¶</a></h2>
<p>Using then <strong>-type</strong> option, <strong>pairToBed</strong> provides several other overlap requirements for controlling how
overlaps between BEDPE/BAM A and BED B are reported. The examples below illustrate how each
option behaves.</p>
<p><strong>-type both</strong>: Report A only if <em>both</em> ends overlap B.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                          ^^^^^^

Result



BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                   ^^^^^^

Result              =====.................................=====</pre>
</div>
<p><strong>-type neither</strong>: Report A only if <em>neither</em> end overlaps B.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                          ^^^^^^

Result



BEDPE/BAM A         *****.................................*****

BED File B   ^^^^                                                  ^^^^^^

Result              =====.................................=====</pre>
</div>
<p><strong>-type xor</strong>: Report A only if <em>one and only one</em> end overlaps B.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                          ^^^^^^

Result              =====.................................=====



BEDPE/BAM A         *****.................................*****

BED File B         ^^^^                                   ^^^^^^

Result</pre>
</div>
<p><strong>-type notboth</strong>: Report A only if <em>neither end</em> <strong>or</strong> <em>one and only one</em> end overlaps B. Thus &#8220;notboth&#8221;
includes what would be reported by &#8220;neither&#8221; and by &#8220;xor&#8221;.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^^^^^^^                                          ^^^^^^

Result              =====.................................=====



BEDPE/BAM A         *****.................................*****

BED File B     ^^^                                               ^^^^^^

Result              =====.................................=====



BEDPE/BAM A         *****.................................*****

BED File B         ^^^^                                   ^^^^^^

Result</pre>
</div>
<p><strong>-type ispan</strong>: Report A if it&#8217;s &#8220;<em>inner span</em>&#8221; overlaps B. Applicable only to intra-chromosomal features.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

              Inner span |-------------------------------|

BEDPE/BAM A         *****.................................*****

BED File B                         ^^^^^^^^

Result              =====.................................=====



BEDPE/BAM A         =====.................................=====

BED File B         ====

Result</pre>
</div>
<p><strong>-type ospan</strong>: Report A if it&#8217;s &#8220;<em>outer span</em>&#8221; overlaps B. Applicable only to intra-chromosomal features.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

        Outer span  |-----------------------------------------|

BEDPE/BAM A         *****.................................*****

BED File B             ^^^^^^^^^^^^

Result              =====.................................=====



BEDPE/BAM A         *****.................................*****

BED File B     ^^^^

Result</pre>
</div>
<p><strong>-type notispan</strong>: Report A only if it&#8217;s &#8220;<em>inner span</em>&#8221; does not overlap B. Applicable only to intrachromosomal
features.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

              Inner span |-------------------------------|

BEDPE/BAM A         *****.................................*****

BED File B                         ^^^^^^^^

Result



BEDPE/BAM A         *****.................................*****

BED File B         ^^^^

Result              =====.................................=====</pre>
</div>
<p><strong>-type notospan</strong>: Report A if it&#8217;s &#8220;<em>outer span</em>&#8221; overlaps B. Applicable only to intra-chromosomal
features.</p>
<div class="highlight-python"><pre>Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

        Outer span  |-----------------------------------------|

BEDPE/BAM A         *****.................................*****

BED File B             ^^^^^^^^^^^^

Result



BEDPE/BAM A         *****.................................*****

BED File B     ^^^^

Result              =====.................................=====</pre>
</div>
</div>
<div class="section" id="f-requiring-a-minimum-overlap-fraction">
<h2>5.2.4 (-f)Requiring a minimum overlap fraction<a class="headerlink" href="#f-requiring-a-minimum-overlap-fraction" title="Permalink to this headline">¶</a></h2>
<p>By default, <strong>pairToBed</strong> will report an overlap between A and B so long as there is at least one base
pair is overlapping on either end. Yet sometimes you may want to restrict reported overlaps between A
and B to cases where the feature in B overlaps at least X% (e.g. 50%) of A. The <strong>?Cf</strong> option does exactly
this. The <strong>-f</strong> option may also be combined with the -type option for additional control. For example,
combining <strong>-f 0.50</strong> with <strong>-type both</strong> requires that both ends of A have at least 50% overlap with a
feature in B.</p>
<p>For example, report A only at least 50% of one of the two ends is overlapped by B.</p>
<div class="highlight-python"><pre>pairToBed -a A.bedpe -b B.bed -f 0.5


Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         *****.................................*****

BED File B         ^^                                           ^^^^^^

Result



BEDPE/BAM A         *****.................................*****

BED File B         ^^^^                                         ^^^^^^

Result              =====.................................=====</pre>
</div>
</div>
<div class="section" id="s-enforcing-strandedness">
<h2>5.2.5 (-s)Enforcing &#8220;strandedness&#8221;<a class="headerlink" href="#s-enforcing-strandedness" title="Permalink to this headline">¶</a></h2>
<p>By default, <strong>pairToBed</strong> will report overlaps between features even if the features are on opposing
strands. However, if strand information is present in both files and the <strong>&#8220;-s&#8221;</strong> option is used, overlaps will
only be reported when features are on the same strand.</p>
<p>For example, report A only at least 50% of one of the two ends is overlapped by B.</p>
<div class="highlight-python"><pre>pairToBed -a A.bedpe -b B.bed -s



Chromosome  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BEDPE/BAM A         &gt;&gt;&gt;&gt;&gt;.................................&lt;&lt;&lt;&lt;&lt;

BED File B         &lt;&lt;                                           &gt;&gt;&gt;&gt;&gt;

Result



BEDPE/BAM A         &gt;&gt;&gt;&gt;&gt;.................................&lt;&lt;&lt;&lt;&lt;

BED File B         &gt;&gt;                                          &gt;&gt;&gt;&gt;&gt;

Result              &gt;&gt;&gt;&gt;&gt;.................................&lt;&lt;&lt;&lt;&lt;</pre>
</div>
</div>
<div class="section" id="abam-default-is-to-write-bam-output-when-using-bam-input">
<h2>5.2.6 (-abam)Default is to write BAM output when using BAM input<a class="headerlink" href="#abam-default-is-to-write-bam-output-when-using-bam-input" title="Permalink to this headline">¶</a></h2>
<p>When comparing <em>paired</em> alignments in BAM format (<strong>-abam</strong>) to features in BED format (<strong>-b</strong>),
<strong>pairToBed</strong> will , by default, write the output in BAM format. That is, each alignment in the BAM
file that meets the user&#8217;s criteria will be written (to standard output) in BAM format. This serves as a
mechanism to create subsets of BAM alignments are of biological interest, etc. Note that both
alignments for each aligned pair will be written to the BAM output.</p>
<p>For example:</p>
<div class="highlight-python"><pre>pairToBed ?Cabam pairedReads.bam ?Cb simreps.bed | samtools view - | head -4

JOBU_0001:3:1:4:1060#0 99 chr10 42387928 29 50M = 42393091 5 2 1 3
AA A A A C G G A A T T A T C G A A T G G A A T C G A A G A G A A T C T T C G A A C G G A C C C G A
dcgggggfbgfgdgggggggfdfgggcggggfcggcggggggagfgbggc XT:A:R NM:i:5 SM:i:0 AM:i:0 X0:i:3 X 1 : i :
3 XM:i:5 XO:i:0 XG:i:0 MD:Z:0T0C33A5T4T3
JOBU_0001:3:1:4:1060#0 147 chr10 42393091 0 50M = 42387928 - 5 2 1 3
AAATGGAATCGAATGGAATCAACATCAAATGGAATCAAATGGAATCATTG K g d c g g d e c d g
\d`ggfcgcggffcgggc^cgfgccgggfc^gcdgg\bg XT:A:R NM:i:2 SM:i:0 AM:i:0 X0:i:3 X1:i:13 XM:i:2 X O : i :
0 XG:i:0 MD:Z:21T14G13
JOBU_0001:3:1:8:446#0 99 chr10 42388091 9 50M = 42392738 4 6 9 7
GAATCGACTGGAATCATCATCGGATGGAAATGAATGGAATAATCATCGAA f _ O f f ` ] I e Y f f ` f f e d d c f e f c P ` c _ W \ \ R _ ]
_BBBBBBBBBBBBBBBB XT:A:U NM:i:4 SM:i:0 AM:i:0 X0:i:1 X1:i:3 XM:i:4 XO:i:0 XG:i:0 M D : Z :
7A22C9C2T6
JOBU_0001:3:1:8:446#0 147 chr10 42392738 9 50M = 42388091 - 4 6 9 7
TTATCGAATGCAATCGAATGGAATTATCGAATGCAATCGAATAGAATCAT df^ffec_JW[`MWceRec``fee`dcecfeeZae`c]
f^cNeecfccf^ XT:A:R NM:i:1 SM:i:0 AM:i:0 X0:i:2 X1:i:2 XM:i:1 XO:i:0 XG:i:0 MD:Z:38A11</pre>
</div>
</div>
<div class="section" id="bedpe-output-bedpe-format-when-using-bam-input">
<h2>5.2.7 (-bedpe)Output BEDPE format when using BAM input<a class="headerlink" href="#bedpe-output-bedpe-format-when-using-bam-input" title="Permalink to this headline">¶</a></h2>
<p>When comparing <em>paired</em> alignments in BAM format (<strong>-abam</strong>) to features in BED format (<strong>-b</strong>),
<strong>pairToBed</strong> will optionally write the output in BEDPE format. That is, each alignment in the BAM
file is converted to a 10 column BEDPE feature and if overlaps are found (or not) based on the user&#8217;s
criteria, the BAM alignment will be reported in BEDPE format. The BEDPE &#8220;name&#8221; field is comprised
of the RNAME field in the BAM alignment. The &#8220;score&#8221; field is the mapping quality score from the
BAM alignment.</p>
<p>For example:</p>
<div class="highlight-python"><pre>pairToBed ?Cabam pairedReads.bam ?Cb simreps.bed -bedpe | head -5
chr10 42387927     42387977    chr10   42393090   42393140
      JOBU_0001:3:1:4:1060#0   29      +     -
chr10 42388090 42388140        chr10   42392737   42392787
      JOBU_0001:3:1:8:446#0    9       +     -
chr10 42390552 42390602        chr10   42396045   42396095
      JOBU_0001:3:1:10:1865#0  9       +     -
chrX  139153741 139153791      chrX    139159018  139159068
      JOBU_0001:3:1:14:225#0   37      +     -
chr4  9236903 9236953          chr4    9242032    9242082
      JOBU_0001:3:1:15:1362#0  0       +     -</pre>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/bedtools.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">5.2 pairToBed</a><ul>
<li><a class="reference internal" href="#usage-and-option-summary">5.2.1 Usage and option summary</a></li>
<li><a class="reference internal" href="#default-behavior">5.2.2 Default behavior</a></li>
<li><a class="reference internal" href="#type-optional-overlap-requirements">5.2.3 (-type)Optional overlap requirements</a></li>
<li><a class="reference internal" href="#f-requiring-a-minimum-overlap-fraction">5.2.4 (-f)Requiring a minimum overlap fraction</a></li>
<li><a class="reference internal" href="#s-enforcing-strandedness">5.2.5 (-s)Enforcing &#8220;strandedness&#8221;</a></li>
<li><a class="reference internal" href="#abam-default-is-to-write-bam-output-when-using-bam-input">5.2.6 (-abam)Default is to write BAM output when using BAM input</a></li>
<li><a class="reference internal" href="#bedpe-output-bedpe-format-when-using-bam-input">5.2.7 (-bedpe)Output BEDPE format when using BAM input</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="intersectBed.html"
                        title="previous chapter">5.1 intersect</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pairToPair.html"
                        title="next chapter">5.3 pairToPair</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/pairToBed.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="pairToPair.html" title="5.3 pairToPair"
             >next</a> |</li>
        <li class="right" >
          <a href="intersectBed.html" title="5.1 intersect"
             >previous</a> |</li>
        <li><a href="../index.html">bedtools v2.16.2</a> &raquo;</li>
          <li><a href="bedtools-suite.html" >The BEDTools suite</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012.
      Last updated on Jul 23, 2012.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>
  </body>
</html>